Language/Variables/Variable Scope & Qualifiers/volatile.adoc

```diff
@@ -55,34 +55,48 @@ There are several ways to do this:
 === Example Code
 // Describe what the example code is all about and add relevant code   ►►►►► THIS SECTION IS MANDATORY ◄◄◄◄◄
 
+The `volatile` modifier ensures that changes to the `state` variable are immediately visible in `loop()`. Without the `volatile` modifier, the `state` variable would be loaded into a register when entering the function and would not be updated anymore until the function ends.
 
 [source,arduino]
 ----
-// toggles LED when interrupt pin changes state
+// Flashes the LED for 1 s if the input has changed
+// in the previous second.
 
-int pin = 13;
 volatile byte state = LOW;
 
 void setup() {
-  pinMode(pin, OUTPUT);
-  attachInterrupt(digitalPinToInterrupt(2), blink, CHANGE);
+  pinMode(LED_BUILTIN, OUTPUT);
+  attachInterrupt(digitalPinToInterrupt(2), toggle, CHANGE);
 }
 
 void loop() {
-  digitalWrite(pin, state);
+  bool changedInTheMeantime = false;
+
+  byte originalState = state;
+  delay(1000);
+  byte newState = state;
+
+  if (newState != originalState) {
+    // toggle() has been called during delay(1000)
+    changedInTheMeantime = true;
+  }
+
+  digitalWrite(LED_BUILTIN, changedInTheMeantime ? HIGH : LOW);
```

Comment:
The interrupt may randomly happen at any time: if it happens outside the delay (for example before line 75 or after line 77), then `originalState` and `newState` will get both the new value of `state` and the change may goes undetected. This is rare I know, but still not correct.

Also the loop looks overly complicated to me, I'd like more something like:

```c++
volatile byte changed = 0;
￼
void loop() {
  if (changed == 1) {
    // toggle() has been called from interrupts!

    // reset changed to 0
    changed = 0;

    // Blink LED for 200 ms
    digitalWrite(LED_BUILTIN, HIGH);
    delay(200);
    digitalWrite(LED_BUILTIN, LOW);
  }
}

void toggle() {
  changed = 1;
}
```